gateway/fields_test.go

```diff
@@ -104,18 +104,99 @@ func TestRetainSingleResult(t *testing.T) {
 	if !reflect.DeepEqual(indata, expdata) {
 		t.Errorf("Unexpected result %v while expecting %v", indata, expdata)
 	}
+}
+
+func TestRetainArray(t *testing.T) {
+	data := `
+	{
+		"result": [
+		  {
+			"x": "1",
+			"y": [
+				{
+				 "a": "2", 
+                 "b": "3"
+				},
+				{
+				 "a": "22", 
+                 "b": "33"
+				}
+			 ]
+		  },
+		  {
+			"x": "4",
+			"y": [
+				{
+				 "a": "5", 
+                 "b": "6"
+				}
+			 ],
+			"z": "5"
+		  }
+		]
+	 }`
+
+	expected := `
+	 {
+		 "result": [
+		   {
+			 "y": [
+				{
+				 "b":"3"
+				},
+				 {
+				 "b":"33"
+				 }
+			  ]
+		   },
+		   {
+			 "y": [
+				{
+				 "b":"6"
+				}
+			 ]
+		   }
+		 ]
+	  }`
 
+	var indata map[string]interface{}
+	err := json.Unmarshal([]byte(data), &indata)
+	if err != nil {
+		t.Errorf("Error parsing test input %s", data)
+		return
+	}
+
+	var expdata map[string]interface{}
+	err = json.Unmarshal([]byte(expected), &expdata)
+	if err != nil {
+		t.Errorf("Error parsing test expected result %s", expected)
+		return
+	}
+
+	req, _ := http.NewRequest("GET", "http://example.com?_fields=y.b", nil)
+
+	ctx := context.Background()
+	retainFields(ctx, req, indata)
+
+	if !reflect.DeepEqual(indata, expdata) {
+		t.Errorf("Unexpected result %v while expecting %v", indata, expdata)
+	}
 }
 
 func TestDoRetain(t *testing.T) {
 	data := `
 	{
 		"a":{
-		   "b":{
-			  "c":"ccc",
-			  "d":"ddd",
-			  "x":"xxx"
-		   },
+		   "b":[
+			  {"c":"ccc",
+			   "d":"ddd",
+			   "x":"xxx"
+		      },
+			  {"c":"ccc2",
+			   "d":"ddd2",
+			   "x":"xxx2"
+		      }
+			],
```

Comment:
I see that you changed all test for nested object to arrays. With this we do not have tests for nested objects. Please revert this and add some new tests for use cases with arrays.